index.js

```diff
@@ -0,0 +1,365 @@
+            console.warn('The plugin of scm-router can not be specified for scms setting');
+
+            return;
+        }
+
+        // eslint-disable-next-line global-require, import/no-dynamic-require
+        const ScmPlugin = require(`screwdriver-scm-${plugin}`);
+        const scmPlugin = new ScmPlugin(options);
+        const scmContexts = scmPlugin.getScmContexts();
+        const scmContext = scmContexts[0]; // plugins return only one scmContext
+
+        if (scmContext == null || scmContext.length === 0) {
+            console.warn(`Illegal scmContext for ${plugin} scm plugin`);
+
+            return;
+        }
+
+        if (typeof this.scms[scmContext] === 'object') {
+            console.warn(`Duplicate scm config for ${scmContext}`);
+
+            return;
+        }
+
+        this.scms[scmContext] = scmPlugin;
+    }
+
+    /**
+     * choose webhook scm module
+     * @method chooseWebhookScm
+     * @param  {Object}     headers          The request headers associated with the webhook payload
+     * @param  {Object}     payload          The webhook payload received from the SCM service
+     * @return {Promise}                     scm object
+     */
+    chooseWebhookScm(headers, payload) {
+        return new Promise((resolve, reject) => {
+            // choose a webhook scm module, or null if there is no suitable one
+            async.detectSeries(this.scms, (scm, cb) => {
+                scm.canHandleWebhook(headers, payload)
+                    .then((result) => {
+                        cb(result === false ? null : scm);
+                    }).catch(() => {
+                        cb(null);
+                    });
+            }, (ret) => {
+                if (ret == null) {
+                    reject('there is no suitable webhook module');
+                }
+
+                resolve(ret);
+            });
+        });
+    }
+
+    /**
+     * choose scm module
+     * @method chooseScm
+     * @param  {Object}     config              Configuration
+     * @param  {String}     config.scmContext   Name of scm context
+     * @return {Promise}                        scm object
+     */
+    chooseScm(config) {
+        return new Promise((resolve, reject) => {
+            if (config.scmContext && typeof config.scmContext === 'string') {
+                const scm = this.scms[config.scmContext];
+
+                if (typeof scm !== 'object') {
+                    reject('Not implemented');
+                }
+
+                resolve(scm);
+            } else {
+                reject('Not implemented');
+            }
+        });
+    }
+
+    /**
+     * Process by all scm module
+     * @method allScm
+     * @param  {function(scm)}                  callback to return map
+     * @return {Promise}                        combined callback results
+     */
+    allScm(callback) {
+        return new Promise((resolve, reject) => {
+            async.mapSeries(Object.keys(this.scms), (key, cb) => {
```

Comment:
So this is just looping through all the scm and do a function call with each of them?
Can we do something like this without async so that we don't need to convert the cb to promise?
https://github.com/screwdriver-cd/build-bookend/blob/master/index.js#L104